### ISSUE TITLE
Should setting CC / CXX request the C backend?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#18450]
Today, setting `CC` in the environment causes `chpl` to use the C backend.

For example:

```
CC=clang CXX=clang++ ./util/printchplenv --all
```

```
CHPL_HOST_PLATFORM: darwin
CHPL_HOST_COMPILER: clang
  CHPL_HOST_CC: clang
  CHPL_HOST_CXX: clang++
CHPL_HOST_ARCH: x86_64
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang
  CHPL_TARGET_CC: clang
  CHPL_TARGET_CXX: clang++
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_RE2: none
CHPL_LLVM: system
  CHPL_LLVM_CONFIG: /usr/local/opt/llvm@11/bin/llvm-config
CHPL_AUX_FILESYS: none
CHPL_LIB_PIC: none
CHPL_SANITIZE: none
CHPL_SANITIZE_EXE: none
```

Note that `CHPL_TARGET_COMPILER=clang` (where the default is `CHPL_TARGET_COMPILER=llvm` because `CHPL_LLVM=system`).

I think this is happening because we use `CC` to infer `CHPL_TARGET_COMPILER`. That's reasonable on the face of it, but might not make sense with the LLVM backend. Perhaps the user just wants to control the *host* compiler.

**Comment by bradcray** [comment]
> we use CC to infer CHPL_TARGET_COMPILER. That's reasonable on the face of it, but might not make sense with the LLVM backend.

Perhaps `CC` should only infer the `CHPL_HOST_COMPILER`, and `CHPL_TARGET_COMPILER` should continue to default to:
* `llvm` if enabled in `chpl`
* whatever `CC` is if `CHPL_HOST_PLATFORM==CHPL_TARGET_PLATFORM`
* whatever it would've defaulted to before the CC-related changes otherwise

**Comment by mppf** [comment]
Yeah. At least this behavior is documented (in chplenv.rst "Note that setting ``CC`` and ``CXX`` will impact both the host and target compiler."). But having `CC` cause `CHPL_HOST_COMPILER` (only) to be inferred sounds like it might be a good direction.

**Comment by mppf** [comment]
In addition to https://github.com/chapel-lang/chapel/issues/18450#issuecomment-926130131 -- we have to figure out when setting CC/CXX impacts the compilation command. I am thinking that we should use the CC/CXX value to select the host and target compiler, *except* if CHPL_TARGET_COMPILER=llvm and then we ignore CC/CXX.

The reason for that is I might have something like `CHPL_TARGET_COMPILER=gnu CC=gcc-12 CXX=g++-12 ./util/printchplenv --all` and here I would expect to see the CC/CXX selection impact the path to the target compiler (even if I am cross-compiling, say).

I wonder if, with this information, we could come up with a simpler rule to the above, e.g. "CC and CXX are ignored when the compilation would otherwise default to the LLVM backend".

**Comment by bradcray** [comment]
Remind me, what happens (today, or alternatively, in  your proposal) if I have, say:

```
CC=gcc
CHPL_HOST_COMPILER=clang
CHPL_TARGET_COMPILER=prgenv-cray
```

And remind me:  When we were designing CC / CXX, did we find any standard environment variable names that are used when cross-compiling?  (My vague recollection is that we did find some but that you said that they had different meanings than what we would want them to be, but I'm not finding that conversation with a quick search).

> CC and CXX are ignored when the compilation would otherwise default to the LLVM backend

I think I can get behind that.  Where I'm currently balking is that without the LLVM back-end enabled, it seems vague to me now whether CC/CXX control host, target, or both, where it seems unintuitive to me that it should control both when cross-compiling.  And ditto if LLVM back-end is enabled but I have multiple arguably conflicting settings, like above.

**Comment by mppf** [comment]
Both my branch and main do this:

```
$ CC=gcc CHPL_HOST_COMPILER=clang CHPL_TARGET_COMPILER=cray-prgenv-cray ./util/printchplenv  --all
Warning: No craype-* processor type module was detected, please load the appropriate one if you want any specialization to occur.
machine info: Linux zeus 5.11.0-37-generic #41-Ubuntu SMP Mon Sep 20 16:39:20 UTC 2021 x86_64
CHPL_HOME: /home/mppf/w/main
script location: /home/mppf/w/main/util/chplenv
CHPL_HOST_PLATFORM: linux64
CHPL_HOST_COMPILER: clang *
  CHPL_HOST_CC: gcc
  CHPL_HOST_CXX: clang++
CHPL_HOST_ARCH: x86_64
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: cray-prgenv-cray *
  CHPL_TARGET_CC: gcc
  CHPL_TARGET_CXX: CC
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: none
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_RE2: none
CHPL_LLVM: system
  CHPL_LLVM_CONFIG: llvm-config-11
CHPL_AUX_FILESYS: none
CHPL_LIB_PIC: none
CHPL_SANITIZE: none
CHPL_SANITIZE_EXE: none
```

> And remind me: When we were designing CC / CXX, did we find any standard environment variable names that are used when cross-compiling? (My vague recollection is that we did find some but that you said that they had different meanings than what we would want them to be, but I'm not finding that conversation with a quick search).

Yes, from the PR message for #17923 :

> ... autotools projects might have HOSTCC that has a different
> meaning. (At least autotools arguably a different meaning for host
> than what we use).

and separately I think you mentioned somewhere that autoconf distinguishes between build, host, and target environments.

I don't remember off-hand what HOSTCC means. [This S.O. question](https://stackoverflow.com/questions/8738086/what-is-the-sense-of-the-word-host-in-hostcc-and-configure-host) has some hints though:

> 
> In autoconf the terminology is:
> 
> * **build** is the system on which you are compiling.
> * **host** is the system on which what you are building will run.
> * **target** is what the compiler you are building will compile for, and is irrelevant when not building a compiler.
> 
> In other tools, for example ptxdist, the terminology is however:
> 
> * **host** is the system you are building on.
> * **target** is the system on which what you are building will run.

Where our use of host/target matches the 2nd here. From the same S.O. question, it doesn't sound like HOSTCC is standardized in autotools and CC is just always used as the compiler.

>  Where I'm currently balking is that without the LLVM back-end enabled, it seems vague to me now whether CC/CXX control host, target, or both, where it seems unintuitive to me that it should control both when cross-compiling.

Yeah, I think that's fair. Perhaps simply having an error/warning for using CC/CXX when cross compiling is the way to go. It could suggest using `CHPL_HOST_CC` / `CHPL_TARGET_CC` etc.

**Comment by mppf** [comment]
In an offline discussion with @bradcray we've settled on the proposal that `CC` / `CXX` will be entirely ignored if the user sets (via environment variable or config file) any of the following variables:
```
CHPL_HOST_COMPILER
CHPL_HOST_CC
CHPL_HOST_CXX
CHPL_TARGET_COMPILER
CHPL_TARGET_CC
CHPL_TARGET_CXX
```

This adds to the idea discussed above that "CC and CXX are ignored when the target compilation would otherwise default to the LLVM backend".

**Comment by mppf** [comment]
On an XC, the common no-LLVM configuration is that CHPL_HOST_COMPILER=gnu (the default there) and CHPL_TARGET_COMPILER=cray-prgenv-cray (or whatever). This ends up being a kind of cross compilation even though none of the variables described above are set. In this case, should we use CC and CXX at all, and if so, how? I am thinking that if a compiler is a PrgEnv compiler, then we don't want to try to customize it with CC/CXX. So, I am thinking that the above list also includes "or if the compiler we are considering inferring from CC etc is a PrgEnv compiler".
